File: engines/physics_engines/mujoco/python/tests/test_spatial_algebra.py
"""
Unit tests for spatial algebra module.

Tests spatial vectors, transformations, and inertia operations.
"""

import numpy as np
import pytest
from mujoco_humanoid_golf.spatial_algebra import (
    crf,
    crm,
    inv_xtrans,
    jcalc,
    mci,
    spatial_cross,
    transform_spatial_inertia,
    xlt,
    xrot,
    xtrans,
)


class TestSpatialCrossProducts:
    """Tests for spatial cross product operators."""

    def test_crm_shape(self) -> None:
        """Test CRM returns correct shape."""
        v = np.array([1, 0, 0, 0, 1, 0])
        X = crm(v)
        assert X.shape == (6, 6)

    def test_crf_crm_relationship(self) -> None:
        """Test that crf(v) = -crm(v).T"""
        v = np.array([1, 2, 3, 4, 5, 6])
        X_crf = crf(v)
        X_crm = crm(v)
        np.testing.assert_allclose(X_crf, -X_crm.T, atol=1e-10)

    def test_spatial_cross_motion(self) -> None:
        """Test motion cross product."""
        v1 = np.array([1, 0, 0, 0, 0, 0])  # Pure angular about x
        v2 = np.array([0, 1, 0, 0, 0, 0])  # Pure angular about y
        result = spatial_cross(v1, v2, "motion")
        expected = np.array([0, 0, 1, 0, 0, 0])  # Should be about z
        np.testing.assert_allclose(result, expected, atol=1e-10)

    def test_spatial_cross_force(self) -> None:
        """Test force cross product."""
        # v = [1, 0, 0, 0, 1, 0] = [omega_x=1, omega_y=0, omega_z=0,
        #                           v_x=0, v_y=1, v_z=0]
        # f = [0, 0, 10, 0, 0, 0] = [moment_x=0, moment_y=0, moment_z=10,
        #                            force_x=0, force_y=0, force_z=0]
        v = np.array([1, 0, 0, 0, 1, 0])
        f = np.array([0, 0, 10, 0, 0, 0])
        result = spatial_cross(v, f, "force")
        assert result.shape == (6,)
        # Verify using crf operator directly
        expected = crf(v) @ f
        np.testing.assert_allclose(result, expected, atol=1e-10)

    def test_spatial_cross_invalid_type(self) -> None:
        """Test spatial_cross raises error for invalid type."""
        v = np.array([1, 0, 0, 0, 0, 0])
        u = np.array([0, 1, 0, 0, 0, 0])
        with pytest.raises(ValueError):
            spatial_cross(v, u, "invalid")

    def test_crm_zero_velocity_returns_zero_matrix(self) -> None:
        """Test CRM returns zero matrix for zero velocity."""
        v = np.zeros(6)
        X = crm(v)
        np.testing.assert_allclose(X, np.zeros((6, 6)), atol=1e-10)

    def test_crf_zero_velocity_returns_zero_matrix(self) -> None:
        """Test CRF returns zero matrix for zero velocity."""
        v = np.zeros(6)
        X = crf(v)
        np.testing.assert_allclose(X, np.zeros((6, 6)), atol=1e-10)


class TestSpatialTransforms:
    """Tests for spatial transformations."""

    def test_xrot_block_diagonal(self) -> None:
        """Test XROT has block diagonal structure."""
        theta = np.pi / 2
        E = np.array(
            [
                [np.cos(theta), -np.sin(theta), 0],
                [np.sin(theta), np.cos(theta), 0],
                [0, 0, 1],
            ],
        )
        X = xrot(E)

        # Check off-diagonal blocks are zero
        np.testing.assert_allclose(X[:3, 3:], 0, atol=1e-10)
        np.testing.assert_allclose(X[3:, :3], 0, atol=1e-10)

        # Check diagonal blocks equal E
        np.testing.assert_allclose(X[:3, :3], E, atol=1e-10)
        np.testing.assert_allclose(X[3:, 3:], E, atol=1e-10)

    def test_xlt_structure(self) -> None:
        """Test XLT has correct structure."""
        r = np.array([1, 2, 3])
        X = xlt(r)

        # Top-right should be zero
        np.testing.assert_allclose(X[:3, 3:], 0, atol=1e-10)

        # Diagonal blocks should be identity
        np.testing.assert_allclose(X[:3, :3], np.eye(3), atol=1e-10)
        np.testing.assert_allclose(X[3:, 3:], np.eye(3), atol=1e-10)

    def test_xtrans_inverse(self) -> None:
        """Test XTRANS and INV_XTRANS are inverses."""
        E = np.array([[0, -1, 0], [1, 0, 0], [0, 0, 1]])  # 90° about z
        r = np.array([1, 2, 0])

        X = xtrans(E, r)
        X_inv = inv_xtrans(E, r)

        product = X @ X_inv
        np.testing.assert_allclose(product, np.eye(6), atol=1e-10)

    def test_xtrans_identity(self) -> None:
        """Test XTRANS with identity rotation."""
        E = np.eye(3)
        r = np.array([1, 2, 3])
        X = xtrans(E, r)

        # Should match xlt
        X_expected = xlt(r)
        np.testing.assert_allclose(X, X_expected, atol=1e-10)

    def test_xrot_invalid_matrix(self) -> None:
        """Test XROT raises error for non-rotation matrix."""
        E = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])  # Not a rotation
        with pytest.raises(ValueError):
            xrot(E)

    def test_xlt_zero_translation(self) -> None:
        """Test XLT with zero translation."""
        r = np.zeros(3)
        X = xlt(r)
        np.testing.assert_allclose(X, np.eye(6), atol=1e-10)


class TestSpatialInertia:
    """Tests for spatial inertia operations."""

    def test_mci_symmetry(self) -> None:
        """Test spatial inertia is symmetric."""
        mass = 1.0
        radius = 0.1
        com = np.array([0, 0, 0])
        I_sphere = (2 / 5) * mass * radius**2 * np.eye(3)

        I_spatial = mci(mass, com, I_sphere)

        # Check symmetry
        np.testing.assert_allclose(I_spatial, I_spatial.T, atol=1e-10)

    def test_mci_positive_definite(self) -> None:
        """Test spatial inertia is positive definite."""
        mass = 2.0
        com = np.array([0.1, 0, 0])
        I_com = 0.01 * np.eye(3)

        I_spatial = mci(mass, com, I_com)

        # All eigenvalues should be positive
        eigvals = np.linalg.eigvals(I_spatial)
        assert np.all(eigvals > 0)

    def test_transform_spatial_inertia_identity(self) -> None:
        """Test identity transform preserves inertia."""
        I_B = mci(1.0, np.zeros(3), 0.01 * np.eye(3))
        X = np.eye(6)
        I_A = transform_spatial_inertia(I_B, X)

        np.testing.assert_allclose(I_A, I_B, atol=1e-10)

    def test_mci_with_offset_com(self) -> None:
        """Test spatial inertia with offset center of mass."""
        mass = 1.0
        com = np.array([0.5, 0, 0])  # COM offset
        I_com = 0.01 * np.eye(3)
        I_spatial = mci(mass, com, I_com)

        # Should still be symmetric and positive definite
        np.testing.assert_allclose(I_spatial, I_spatial.T, atol=1e-10)
        eigvals = np.linalg.eigvals(I_spatial)
        assert np.all(eigvals > 0)

    def test_transform_spatial_inertia_rotation(self) -> None:
        """Test transforming inertia with rotation."""
        I_B = mci(1.0, np.zeros(3), 0.01 * np.eye(3))
        # 90 degree rotation about z
        theta = np.pi / 2
        E = np.array(
            [
                [np.cos(theta), -np.sin(theta), 0],
                [np.sin(theta), np.cos(theta), 0],
                [0, 0, 1],
            ],
        )
        X = xrot(E)
        I_A = transform_spatial_inertia(I_B, X)

        # Should still be symmetric and positive definite
        np.testing.assert_allclose(I_A, I_A.T, atol=1e-10)
        eigvals = np.linalg.eigvals(I_A)
        assert np.all(eigvals > 0)


class TestJointCalculations:
    """Tests for joint kinematics."""

    def test_jcalc_revolute_z(self) -> None:
        """Test revolute joint about z-axis."""
        Xj, S = jcalc("Rz", np.pi / 4)

        # Motion subspace should be [0, 0, 1, 0, 0, 0]
        expected_S = np.array([0, 0, 1, 0, 0, 0])
        np.testing.assert_allclose(S, expected_S, atol=1e-10)

        # Xj should be 6x6
        assert Xj.shape == (6, 6)

    def test_jcalc_prismatic_x(self) -> None:
        """Test prismatic joint along x-axis."""
        _Xj, S = jcalc("Px", 0.5)

        # Motion subspace should be [0, 0, 0, 1, 0, 0]
        expected_S = np.array([0, 0, 0, 1, 0, 0])
        np.testing.assert_allclose(S, expected_S, atol=1e-10)

    def test_jcalc_all_joint_types(self) -> None:
        """Test all supported joint types."""
        joint_types = ["Rx", "Ry", "Rz", "Px", "Py", "Pz"]

        for jtype in joint_types:
            Xj, S = jcalc(jtype, 0.1)
            assert Xj.shape == (6, 6)
            assert S.shape == (6,)

    def test_jcalc_unsupported_joint(self) -> None:
        """Test error for unsupported joint type."""
        with pytest.raises(ValueError):
            jcalc("invalid", 0)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: engines/physics_engines/mujoco/matlab/tests/test_spatial_algebra.m
function test_spatial_algebra()
% TEST_SPATIAL_ALGEBRA  Unit tests for spatial algebra functions
%   Comprehensive tests for spatial_v2 library including:
%   - Spatial cross product operators (CRM, CRF)
%   - Spatial transformations (XROT, XLT, XTRANS)
%   - Spatial inertia (MCI, TRANSFORM_SPATIAL_INERTIA)
%   - Joint calculations (JCALC)

fprintf('Running spatial algebra tests...\n');

% Test counter
num_tests = 0;
num_passed = 0;

% Add paths
addpath('../spatial_v2');
addpath('..');
constants; % Load constants into workspace

%% Test 1: CRM - Spatial cross product operator for motion
num_tests = num_tests + 1;
try
    v = [1; 0; 0; 0; 1; 0];  % Angular [1,0,0], linear [0,1,0]
    X = crm(v);

    % Check dimensions
    assert(all(size(X) == [SPATIAL_DIM, SPATIAL_DIM]), 'CRM: Wrong size');

    % Check structure: top-left and bottom-right should be skew(w)
    assert(abs(X(1,1)) < eps, 'CRM: Diagonal should be zero');

    % Verify antisymmetry of blocks
    w_skew = X(1:3, 1:3);
    assert(norm(w_skew + w_skew', 'fro') < 1e-10, 'CRM: Should be skew-symmetric');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 1: CRM operator\n');
catch ME
    fprintf('  [FAIL] Test 1: CRM operator - %s\n', ME.message);
end

%% Test 2: CRF - Dual spatial cross product operator
num_tests = num_tests + 1;
try
    v = [1; 0; 0; 0; 1; 0];
    X_crf = crf(v);
    X_crm = crm(v);

    % Verify relationship: crf(v) = -crm(v)'
    assert(norm(X_crf + X_crm', 'fro') < 1e-10, ...
        'CRF: Should equal -CRM transpose');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 2: CRF operator\n');
catch ME
    fprintf('  [FAIL] Test 2: CRF operator - %s\n', ME.message);
end

%% Test 3: XROT - Rotation transformation
num_tests = num_tests + 1;
try
    % 90 degree rotation about z-axis
    theta = pi/2;
    E = [cos(theta), -sin(theta), 0;
         sin(theta),  cos(theta), 0;
         0,           0,          1];
    X = xrot(E);

    % Check block diagonal structure
    assert(norm(X(1:SPATIAL_LIN_DIM, SPATIAL_LIN_DIM+1:SPATIAL_DIM), 'fro') < 1e-10, 'XROT: Off-diagonal should be zero');
    assert(norm(X(SPATIAL_LIN_DIM+1:SPATIAL_DIM, 1:SPATIAL_LIN_DIM), 'fro') < 1e-10, 'XROT: Off-diagonal should be zero');
    assert(norm(X(1:SPATIAL_LIN_DIM, 1:SPATIAL_LIN_DIM) - E, 'fro') < 1e-10, 'XROT: Top block wrong');
    assert(norm(X(SPATIAL_LIN_DIM+1:SPATIAL_DIM, SPATIAL_LIN_DIM+1:SPATIAL_DIM) - E, 'fro') < 1e-10, 'XROT: Bottom block wrong');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 3: XROT transformation\n');
catch ME
    fprintf('  [FAIL] Test 3: XROT transformation - %s\n', ME.message);
end

%% Test 4: XLT - Translation transformation
num_tests = num_tests + 1;
try
    r = [1; 2; 3];
    X = xlt(r);

    % Check structure: top-right should be zero, others identity or -r_skew
    assert(norm(X(1:SPATIAL_LIN_DIM, SPATIAL_LIN_DIM+1:SPATIAL_DIM), 'fro') < 1e-10, 'XLT: Top-right should be zero');
    assert(norm(X(1:SPATIAL_LIN_DIM, 1:SPATIAL_LIN_DIM) - eye(SPATIAL_LIN_DIM), 'fro') < 1e-10, 'XLT: Top-left should be I');
    assert(norm(X(SPATIAL_LIN_DIM+1:SPATIAL_DIM, SPATIAL_LIN_DIM+1:SPATIAL_DIM) - eye(SPATIAL_LIN_DIM), 'fro') < 1e-10, 'XLT: Bottom-right should be I');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 4: XLT transformation\n');
catch ME
    fprintf('  [FAIL] Test 4: XLT transformation - %s\n', ME.message);
end

%% Test 5: XTRANS and INV_XTRANS
num_tests = num_tests + 1;
try
    E = [0, -1, 0; 1, 0, 0; 0, 0, 1];  % 90° rotation about z
    r = [1; 2; 0];

    X = xtrans(E, r);
    X_inv = inv_xtrans(E, r);

    % Verify inverse relationship
    prod = X * X_inv;
    assert(norm(prod - eye(SPATIAL_DIM), 'fro') < 1e-10, ...
        'XTRANS: X * X_inv should equal identity');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 5: XTRANS and inverse\n');
catch ME
    fprintf('  [FAIL] Test 5: XTRANS and inverse - %s\n', ME.message);
end

%% Test 6: MCI - Spatial inertia construction
num_tests = num_tests + 1;
try
    % Simple sphere at origin
    mass = 1.0;
    radius = 0.1;
    com = [0; 0; 0];
    I_sphere = (2/5) * mass * radius^2 * eye(3);

    I_spatial = mcI(mass, com, I_sphere);

    % Check symmetry
    assert(norm(I_spatial - I_spatial', 'fro') < 1e-10, ...
        'MCI: Spatial inertia should be symmetric');

    % Check positive definiteness (all eigenvalues positive)
    eigvals = eig(I_spatial);
    assert(all(eigvals > 0), 'MCI: Should be positive definite');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 6: MCI spatial inertia\n');
catch ME
    fprintf('  [FAIL] Test 6: MCI spatial inertia - %s\n', ME.message);
end

%% Test 7: JCALC - Revolute joint
num_tests = num_tests + 1;
try
    [Xj, S] = jcalc('Rz', pi/4);

    % Check motion subspace for Rz joint
    S_expected = [0; 0; 1; 0; 0; 0];
    assert(norm(S - S_expected) < 1e-10, 'JCALC: Wrong motion subspace for Rz');

    % Verify Xj is a valid spatial transform
    assert(all(size(Xj) == [SPATIAL_DIM, SPATIAL_DIM]), 'JCALC: Wrong transform size');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 7: JCALC revolute joint\n');
catch ME
    fprintf('  [FAIL] Test 7: JCALC revolute joint - %s\n', ME.message);
end

%% Test 8: JCALC - Prismatic joint
num_tests = num_tests + 1;
try
    [Xj, S] = jcalc('Px', 0.5);

    % Check motion subspace for Px joint
    S_expected = [0; 0; 0; 1; 0; 0];
    assert(norm(S - S_expected) < 1e-10, 'JCALC: Wrong motion subspace for Px');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 8: JCALC prismatic joint\n');
catch ME
    fprintf('  [FAIL] Test 8: JCALC prismatic joint - %s\n', ME.message);
end

%% Test 9: SPATIAL_CROSS - Motion cross product
num_tests = num_tests + 1;
try
    v1 = [1; 0; 0; 0; 0; 0];  % Pure angular velocity about x
    v2 = [0; 1; 0; 0; 0; 0];  % Pure angular velocity about y

    result = spatial_cross(v1, v2, 'motion');

    % v1 x v2 should give angular velocity about z
    expected = [0; 0; 1; 0; 0; 0];
    assert(norm(result - expected) < 1e-10, ...
        'SPATIAL_CROSS: Wrong result for motion cross product');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 9: SPATIAL_CROSS motion\n');
catch ME
    fprintf('  [FAIL] Test 9: SPATIAL_CROSS motion - %s\n', ME.message);
end

%% Test 10: TRANSFORM_SPATIAL_INERTIA
num_tests = num_tests + 1;
try
    % Create a simple inertia
    mass = 1.0;
    I_com = 0.01 * eye(3);
    com = [0; 0; 0];
    I_B = mcI(mass, com, I_com);

    % Identity transform should not change inertia
    X = eye(SPATIAL_DIM);
    I_A = transform_spatial_inertia(I_B, X);

    assert(norm(I_A - I_B, 'fro') < 1e-10, ...
        'TRANSFORM_SPATIAL_INERTIA: Identity transform should preserve inertia');

    num_passed = num_passed + 1;
    fprintf('  [PASS] Test 10: TRANSFORM_SPATIAL_INERTIA\n');
catch ME
    fprintf('  [FAIL] Test 10: TRANSFORM_SPATIAL_INERTIA - %s\n', ME.message);
end

%% Summary
fprintf('\n=== Test Summary ===\n');
fprintf('Passed: %d/%d tests\n', num_passed, num_tests);
if num_passed == num_tests
    fprintf('All tests PASSED!\n');
else
    fprintf('Some tests FAILED.\n');
end
end
